### PR TITLE
feat: extend dryRun options to extract lookup full

### LINKED ISSUE #14355: Extend dryRun options to extract, lookup, update
### What would you like Renovate to be able to do?

Change `dryRun` option from boolean to enum, supporting:
- `"extract"`: Performs extract only, prints results, exits
- `"lookup"`: Performs extract and lookup, but exits before running through branches
- `"full"`: Same as today's `true` - runs through as far as it can

### If you have any ideas on how this should be implemented, please tell us here.

We need to add migration for the existing option, including in env and cli.

We need to add specific logging after `extract` if we are exiting then, because otherwise the useful `packageFiles with updates` message won't be reached.

We ideally should exit the `lookup` phase before branchifying and flattening, so in that case we should also log the packageFiles with updates message, which will be almost the same as today but without the `branchName`.

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by MaronHatoum**
## Changes
feature: I changed dryRun(optional property) type from boolean to string(allowed values: "full", "lookup", "extract")
the new options have been supported and dryRun can run with them.

## Context
Closes #14355

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by MaronHatoum** [comment]
**Documentation changes needed.**

File: lib/config/options/index.ts [hunk 1]
```diff
@@ -246,9 +246,10 @@ const options: RenovateOptions[] = [
     name: 'dryRun',
     description:
       'If enabled, perform a dry run by logging messages instead of creating/updating/deleting branches and PRs.',
-    type: 'boolean',
+    type: 'string',
     globalOnly: true,
-    default: false,
+    allowedValues: ['extract', 'lookup', 'full'],
+    default: '',
```
Comment (viceice) [inline lib/config/options/index.ts:252]: I think it should be null instead of empty string

**Review by viceice — DISMISSED** [review]
Needs migration code to not break existing users.

File: lib/config/options/index.ts (hunk 1 above)
Comment (MaronHatoum) [reply lib/config/options/index.ts:252]: what you mean with migration code to not break existing users??
if user inserted true or false instead of the new allowed values,
we should support this case also?

Comment (viceice) [reply lib/config/options/index.ts:252]: Yes, we always try to be backwards compatible. So please check migrations in `lib/config` folder for samples

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -245,10 +245,14 @@ const options: RenovateOptions[] = [
   {
     name: 'dryRun',
     description:
-      'If enabled, perform a dry run by logging messages instead of creating/updating/deleting branches and PRs.',
-    type: 'boolean',
+      '"extract": Performs extract only, prints results, exits.\n' +
+      '"lookup": Performs extract and lookup, but exits before running through branches.\n' +
+      '"full": Perform a dry run by logging messages instead of creating/updating/deleting branches and PRs.\n' +
+      'Default value = null',
```
Comment (rarkins) [inline lib/config/options/index.ts:251]: This level of detail should be moved to the markdown doc

File: lib/workers/global/config/parse/index.ts [hunk 3]
```diff
@@ -22,6 +22,19 @@ export async function parseConfigs(
...
+  if (envConfig?.dryRun === 'true') {
+    envConfig.dryRun = 'full';
+  }
+  if (cliConfig?.dryRun === 'false') {
+    cliConfig.dryRun = null;
+  }
+  if (envConfig?.dryRun === 'false') {
+    envConfig.dryRun = null;
+  }
+
```
Comment (rarkins) [inline lib/workers/global/config/parse/index.ts:37]: I think we should log a warning here that the field values have changed, so that we can remove this code in a next major release

File: lib/workers/repository/dependency-dashboard.ts [hunk 4]
```diff
@@ -117,7 +117,7 @@ export async function ensureDependencyDashboard(
       )
     )
   ) {
-    if (GlobalConfig.get('dryRun')) {
+    if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/dependency-dashboard.ts:120]: I think if we got this far then you no longer need to check for the value of `dryRun` - just the presence of it. So all these `=== 'full'` additions could be removed.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts (hunk 2 above)
Comment (MaronHatoum) [reply lib/config/options/index.ts:251]: do you mean it should appear under the table?

Comment (MaronHatoum) [reply lib/config/options/index.ts:251]: something like this?
image

File: lib/workers/repository/error-config.ts [hunk 5]
```diff
@@ -26,7 +26,7 @@ export async function raiseConfigWarningIssue(
     logger.debug('Updating onboarding PR with config error notice');
     body = `## Action Required: Fix Renovate Configuration\n\n${body}`;
     body += `\n\nOnce you have resolved this problem (in this onboarding branch), Renovate will return to providing you with a preview of your repository's configuration.`;
-    if (GlobalConfig.get('dryRun')) {
+    if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/error-config.ts:29]:
```suggestion
    if (GlobalConfig.get('dryRun')) {
```

We don't want to do this under any dry run scenario

File: lib/workers/repository/error-config.ts [hunk 6]
```diff
@@ -39,7 +39,7 @@ export async function raiseConfigWarningIssue(
         logger.warn({ err }, 'Error updating onboarding PR');
       }
     }
-  } else if (GlobalConfig.get('dryRun')) {
+  } else if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/error-config.ts:42]:
```suggestion
  } else if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/finalise/prune.ts [hunk 7]
```diff
@@ -32,7 +32,7 @@ async function cleanUpBranches(
             { prNo: pr.number, prTitle: pr.title },
             'Branch is modified - skipping PR autoclosing'
           );
-          if (GlobalConfig.get('dryRun')) {
+          if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/finalise/prune.ts:35]:
```suggestion
          if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/finalise/prune.ts [hunk 8]
```diff
@@ -63,9 +63,9 @@ async function cleanUpBranches(
           });
           await deleteBranch(branchName);
         }
-      } else if (GlobalConfig.get('dryRun')) {
+      } else if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/finalise/prune.ts:66]:
```suggestion
      } else if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/finalise/prune.ts [hunk 9]
```diff
@@ -63,9 +63,9 @@ async function cleanUpBranches(
...
         logger.info(`DRY-RUN: Would delete orphan branch ${branchName}`);
-      } else {
+      } else if (!GlobalConfig.get('dryRun')) {
```
Comment (rarkins) [inline lib/workers/repository/finalise/prune.ts:68]:
```suggestion
      } else {
```

File: lib/workers/repository/index.ts [hunk 10]
```diff
@@ -54,7 +54,10 @@ export async function renovateRepository(
         return recursiveRes;
       }
       logger.debug(`Automerged but already retried once`);
-    } else {
+    } else if (
+      GlobalConfig.get('dryRun') !== 'lookup' &&
+      GlobalConfig.get('dryRun') !== 'extract'
+    ) {
```
Comment (rarkins) [inline lib/workers/repository/index.ts:60]: Does this function already have its own dry run check? I'm not sure this code is necessary

File: lib/workers/repository/onboarding/branch/create.ts [hunk 11]
```diff
@@ -26,7 +26,7 @@ export async function createOnboardingBranch(
   const commitMessage = commitMessageFactory.create();
 
   // istanbul ignore if
-  if (GlobalConfig.get('dryRun')) {
+  if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/onboarding/branch/create.ts:29]:
```suggestion
  if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 12]
```diff
@@ -39,7 +39,7 @@ export async function rebaseOnboardingBranch(
   const commitMessage = commitMessageFactory.create();
 
   // istanbul ignore if
-  if (GlobalConfig.get('dryRun')) {
+  if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/onboarding/branch/rebase.ts:42]:
```suggestion
  if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/onboarding/pr/index.ts [hunk 13]
```diff
@@ -76,7 +76,7 @@ If you need any further assistance then you can also [request help here](${confi
     prBody = prBody.replace('{{PACKAGE FILES}}\n', '');
   }
   let configDesc = '';
-  if (GlobalConfig.get('dryRun')) {
+  if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/onboarding/pr/index.ts:79]:
```suggestion
  if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/onboarding/pr/index.ts [hunk 14]
```diff
@@ -124,7 +124,7 @@ If you need any further assistance then you can also [request help here](${confi
       return;
     }
     // PR must need updating
-    if (GlobalConfig.get('dryRun')) {
+    if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/onboarding/pr/index.ts:127]:
```suggestion
    if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/onboarding/pr/index.ts [hunk 15]
```diff
@@ -139,7 +139,7 @@ If you need any further assistance then you can also [request help here](${confi
   logger.debug('Creating onboarding PR');
   const labels: string[] = prepareLabels(config);
   try {
-    if (GlobalConfig.get('dryRun')) {
+    if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/onboarding/pr/index.ts:142]:
```suggestion
    if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/process/deprecated.ts [hunk 16]
```diff
@@ -55,7 +55,7 @@ export async function raiseDeprecationWarnings(
         .join(', ')}`;
       issueBody += `\n\nIf you don't care about this, you can close this issue and not be warned about \`${depName}\`'s deprecation again. If you would like to completely disable all future deprecation warnings then add the following to your config:\n\n\`\`\`\n"suppressNotifications": ["deprecationWarningIssues"]\n\`\`\`\n\n`;
       // istanbul ignore if
-      if (GlobalConfig.get('dryRun')) {
+      if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/process/deprecated.ts:58]:
```suggestion
      if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/process/index.ts [hunk 17]
```diff
@@ -103,9 +104,11 @@ export async function extractDependencies(
...
     const packageFiles = await extract(config);
     addSplit('extract');
-    res = await lookup(config, packageFiles);
+    if (GlobalConfig.get('dryRun') !== 'extract') {
+      res = await lookup(config, packageFiles);
+      addSplit('lookup');
+    }
   }
-  addSplit('lookup');
   return res;
```
Comment (rarkins) [inline lib/workers/repository/process/index.ts:112]: I'd prefer to revert these changes and instead add logic in the right place for "if dryRun===extract then return"

File: lib/workers/repository/process/write.ts [hunk 18]
```diff
@@ -34,8 +35,15 @@ export async function writeUpdates(
...
-    const res = await processBranch(branch);
+    let branchExisted;
+    let res;
+    if (
+      GlobalConfig.get('dryRun') !== 'extract' &&
+      GlobalConfig.get('dryRun') !== 'lookup'
+    ) {
+      branchExisted = branchExists(branch.branchName);
+      res = await processBranch(branch);
+    }
```
Comment (rarkins) [inline lib/workers/repository/process/write.ts:46]: Again here, it would be better to add a "if dryRun===extract|lookup then return" before doing any branch work

File: lib/workers/repository/update/branch/artifacts.ts [hunk 19]
```diff
@@ -23,7 +23,7 @@ export async function setArtifactErrorStatus(
   // Check if state needs setting
   if (existingState !== state) {
     logger.debug(`Updating status check state to failed`);
-    if (GlobalConfig.get('dryRun')) {
+    if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/branch/artifacts.ts:26]: revert

File: lib/workers/repository/update/branch/automerge.ts [hunk 20]
```diff
@@ -33,7 +33,7 @@ export async function tryBranchAutomerge(
   if (branchStatus === BranchStatus.green) {
     logger.debug(`Automerging branch`);
     try {
-      if (GlobalConfig.get('dryRun')) {
+      if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/branch/automerge.ts:36]: revert

File: lib/workers/repository/update/branch/handle-existing.ts [hunk 21]
```diff
@@ -20,7 +20,7 @@ export async function handlepr(config: BranchConfig, pr: Pr): Promise<void> {
     content +=
       '\n\nIf this PR was closed by mistake or you changed your mind, you can simply rename this PR and you will soon get a fresh replacement PR opened.';
     if (!config.suppressNotifications.includes('prIgnoreNotification')) {
-      if (GlobalConfig.get('dryRun')) {
+      if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/branch/handle-existing.ts:23]: revert all changes in this file

File: lib/workers/repository/update/branch/index.ts [hunk 22]
```diff
@@ -431,11 +431,11 @@ export async function processBranch(
     } else if (config.updatedArtifacts?.length && branchPr) {
       // If there are artifacts, no errors, and an existing PR then ensure any artifacts error comment is removed
       // istanbul ignore if
-      if (GlobalConfig.get('dryRun')) {
+      if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/branch/index.ts:434]: revert all changes in this file

File: lib/workers/repository/update/branch/reuse.ts [hunk 23]
```diff
@@ -43,7 +43,7 @@ export async function shouldReuseExistingBranch(
     if (pr.labels?.includes(config.rebaseLabel)) {
       logger.debug(`Manual rebase requested via PR labels for #${pr.number}`);
       // istanbul ignore if
-      if (GlobalConfig.get('dryRun')) {
+      if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/branch/reuse.ts:46]: revert

**Review by rarkins — CHANGES_REQUESTED** [review]
Nearly all checks for `if dryRun === 'full'` are either incorrect or unnecessary:
- They are mostly unnecessary because if dryRun===extract or lookup then the code block shouldn't be reachable in the first place, or
- If the code is reachable, perhaps due to a bug, then we still want to treat it as a dry run and NOT do updates

File: lib/workers/repository/index.ts (hunk 10 above)
Comment (MaronHatoum) [reply lib/workers/repository/index.ts:60]: I added this condition because we don't need to access the `ensureDependencyDashboard(config, branches)` when running with extract or lookup mode.

Comment (MaronHatoum) [reply lib/workers/repository/index.ts:60]: you are right this is not necessary because in case we are running with extract or lookup mode the flow should be ended before this point.
I will delete this "if"

File: docs/usage/self-hosted-configuration.md [hunk 24]
```diff
@@ -292,6 +292,11 @@ Like this:
 
 ## dryRun
 
+"extract": Performs extract only, prints results, exits.<br>
+"lookup": Performs extract and lookup, but exits before running through branches.<br>
+"full": Perform a dry run by logging messages instead of creating/updating/deleting branches and PRs.<br>
+Default value = null.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:298]:
```suggestion
You can choose from the following behaviors for the `dryRun` config option:

- `null`: REPLACE WITH EXPLANATION, this is the default value
- `"extract"`: Performs extract only, prints results, exits
- `"lookup"`: Performs extract and lookup, but exits before running through branches
- `"full"`: Performs a dry run by logging messages instead of creating/updating/deleting branches and PRs.
```

Changes:

- Convert to list
  - Make `null` default the top list item
- Small rewrite so that the list items start with `performs`
- Add `monospaced` font to the config option strings
- Remove `<br>` HTML code

After applying my suggestion, please replace the `REPLACE WITH EXPLANATION` text with a proper explanation of what `null` means. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md [hunk 25]
```diff
@@ -292,6 +292,13 @@ Like this:
 
 ## dryRun
 
+You can choose from the following behaviors for the `dryRun` config option:
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:295]: Can you add a sentence above this one that explains what a "dry run" is, and why you might want to use it?

I guess it's a way to "preview" the behavior of Renovate, which doesn't actually make any changes to the repository files?

File: docs/usage/self-hosted-configuration.md [hunk 26]
```diff
@@ -292,6 +292,13 @@ Like this:
...
+
+- `null`: Performs a regular renovate run include creating/updating/deleting branches and PRs, this is the default value
+- `"extract"`: Performs extract only, prints results, exits
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:298]: I'm not sure what "extract only" means here, do we need to explain that in more detail here?

File: docs/usage/self-hosted-configuration.md [hunk 27]
```diff
@@ -292,6 +292,13 @@ Like this:
...
+- `"lookup"`: Performs extract and lookup, but exits before running through branches
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:299]: Do we need to explain the lookup as well?

File: docs/usage/self-hosted-configuration.md [hunk 28]
```diff
@@ -292,6 +292,13 @@ Like this:
...
+- `"full"`: Performs a dry run by logging messages instead of creating/updating/deleting branches and PRs.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:300]:
```suggestion
- `"full"`: Performs a dry run by logging messages instead of creating/updating/deleting branches and PRs
```

We don't use punctuation at the end of list items.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
I don't know what these terms mean when you're doing a "dry run":

- extract
- lookup
- logging

Do we want to add explanation here, or put it somewhere else?

File: docs/usage/self-hosted-configuration.md [hunk 29]
```diff
@@ -292,12 +292,14 @@ Like this:
 
 ## dryRun
 
+`dryRun` Preview the behavior of Renovate in logs, without making any changes to the repository files
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:295]:
```suggestion
`dryRun` Preview the behavior of Renovate in logs, without making any changes to the repository files.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md [hunk 30]
```diff
@@ -292,6 +292,15 @@ Like this:
 
 ## dryRun
 
+`dryRun` Preview the behavior of Renovate in logs, without making any changes to the repository files.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:295]:
```suggestion
Use `dryRun` to preview the behavior of Renovate in logs, without making any changes to the repository files.
```

File: lib/workers/global/config/parse/index.ts [hunk 31]
```diff
@@ -22,6 +22,27 @@ export async function parseConfigs(
...
+    logger.warn('cli config dryRun property has been changed to full');
+    cliConfig.dryRun = 'full';
+  }
+  if (envConfig?.dryRun === 'true') {
+    logger.warn('env config dryRun property has been changed to full');
+    envConfig.dryRun = 'full';
+  }
+  if (cliConfig?.dryRun === 'false' || cliConfig?.dryRun === 'null') {
+    logger.warn(
+      'cli config dryRun property has been changed to null, running with normal mood.'
```
Comment (HonkingGoose) [inline lib/workers/global/config/parse/index.ts:35]:
```suggestion
      'cli config dryRun property has been changed to null, running with normal mode.'
```

I think you mean `normal mode` here. 😉

But maybe Renovate bot can really get moody, and be grumpy about it's working hours/conditions. Maybe it needs a break. 🤣

File: lib/workers/global/config/parse/index.ts [hunk 32]
```diff
@@ -22,6 +22,27 @@ export async function parseConfigs(
...
+    );
+    cliConfig.dryRun = null;
+  }
+  if (envConfig?.dryRun === 'false' || envConfig?.dryRun === 'null') {
+    logger.warn(
+      'env config dryRun property has been changed to null, running with normal mood.'
```
Comment (HonkingGoose) [inline lib/workers/global/config/parse/index.ts:41]:
```suggestion
      'env config dryRun property has been changed to null, running with normal mode.'
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/workers/global/config/parse/index.ts (hunk 31 above)
Comment (MaronHatoum) [reply lib/workers/global/config/parse/index.ts:35]: 🤣🤣🤣

File: lib/workers/repository/finalise/prune.ts [hunk 33]
```diff
@@ -42,7 +42,7 @@ async function cleanUpBranches(
                 'This PR has been flagged for autoclosing, however it is being skipped due to the branch being already modified. Please close/delete it manually or report a bug if you think this is in error.',
             });
           }
-        } else if (GlobalConfig.get('dryRun')) {
+        } else if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/finalise/prune.ts:45]:
```suggestion
        } else if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/process/write.ts [hunk 34]
```diff
@@ -34,6 +35,12 @@ export async function writeUpdates(
 
   for (const branch of branches) {
     addMeta({ branch: branch.branchName });
+    if (
+      GlobalConfig.get('dryRun') === 'extract' ||
+      GlobalConfig.get('dryRun') === 'lookup'
+    ) {
+      return 'done';
+    }
```
Comment (rarkins) [inline lib/workers/repository/process/write.ts:43]: I don't think we should let it get this far if dryRun is extract or lookup

File: lib/workers/repository/update/branch/commit.ts [hunk 35]
```diff
@@ -31,7 +31,7 @@ export function commitFilesToBranch(
   const fileLength = [...new Set(updatedFiles.map((file) => file.path))].length;
   logger.debug(`${fileLength} file(s) to commit`);
   // istanbul ignore if
-  if (GlobalConfig.get('dryRun')) {
+  if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/branch/commit.ts:34]:
```suggestion
  if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/update/pr/automerge.ts [hunk 36]
```diff
@@ -87,7 +87,7 @@ export async function checkAutoMerge(
   if (automergeType === 'pr-comment') {
     logger.debug(`Applying automerge comment: ${automergeComment}`);
     // istanbul ignore if
-    if (GlobalConfig.get('dryRun')) {
+    if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/pr/automerge.ts:90]:
```suggestion
    if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/update/pr/automerge.ts [hunk 37]
```diff
@@ -96,6 +96,12 @@ export async function checkAutoMerge(
         prAutomergeBlockReason: PrAutomergeBlockReason.DryRun,
       };
     }
+    if (GlobalConfig.get('dryRun') === 'lookup') {
+      return {
+        automerged: false,
+        prAutomergeBlockReason: PrAutomergeBlockReason.DryRun,
+      };
+    }
```
Comment (rarkins) [inline lib/workers/repository/update/pr/automerge.ts:104]:
```suggestion
```

It shouldn't get this far, and needs to be fixed earlier in the logic flow if it does

File: lib/workers/repository/update/pr/automerge.ts [hunk 38]
```diff
@@ -112,7 +118,7 @@ export async function checkAutoMerge(
   }
   // Let's merge this
   // istanbul ignore if
-  if (GlobalConfig.get('dryRun')) {
+  if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/pr/automerge.ts:121]:
```suggestion
  if (GlobalConfig.get('dryRun')) {
```

File: lib/workers/repository/update/pr/automerge.ts [hunk 39]
```diff
@@ -121,6 +127,12 @@ export async function checkAutoMerge(
       prAutomergeBlockReason: PrAutomergeBlockReason.DryRun,
     };
   }
+  if (GlobalConfig.get('dryRun') === 'lookup') {
+    return {
+      automerged: false,
+      prAutomergeBlockReason: PrAutomergeBlockReason.DryRun,
+    };
+  }
```
Comment (rarkins) [inline lib/workers/repository/update/pr/automerge.ts:135]:
```suggestion
```

If dry-run is lookup, we should never be in the "update" logic here in the first place

File: lib/workers/repository/update/pr/index.ts [hunk 40]
```diff
@@ -85,7 +85,7 @@ export async function addAssigneesReviewers(
       }
       if (assignees.length > 0) {
         // istanbul ignore if
-        if (GlobalConfig.get('dryRun')) {
+        if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:88]: revert

File: lib/workers/repository/update/pr/index.ts [hunk 41]
```diff
@@ -114,7 +114,7 @@ export async function addAssigneesReviewers(
       }
       if (reviewers.length > 0) {
         // istanbul ignore if
-        if (GlobalConfig.get('dryRun')) {
+        if (GlobalConfig.get('dryRun') === 'full') {
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:117]: revert EVERY check like this in "update" logic

File: lib/workers/repository/update/pr/index.ts [hunk 42]
```diff
@@ -151,6 +151,7 @@ export function getPlatformPrOptions(
 export type ResultWithPr = {
   type: 'with-pr';
   pr: Pr;
+  prBlockedBy?: never;
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:154]: Is this related/necessary for this PR?

**Review by rarkins — CHANGES_REQUESTED** [review]
Pretty much the same changes as requested before

File: lib/workers/repository/update/pr/index.ts [hunk 43]
```diff
@@ -413,7 +413,7 @@ export async function ensurePr(
       // istanbul ignore if
       if (GlobalConfig.get('dryRun')) {
         logger.info(`DRY-RUN: Would update PR #${existingPr.number}`);
-      } else {
+      } else if (!GlobalConfig.get('dryRun')) {
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:416]: I think these added if statements in this file can be reverted? There's an if before them that checks if dryRun is set so these can't possibly be false

File: lib/workers/repository/update/pr/index.ts [hunk 44]
```diff
@@ -435,7 +435,7 @@ export async function ensurePr(
       if (GlobalConfig.get('dryRun')) {
         logger.info('DRY-RUN: Would create PR: ' + prTitle);
         pr = { number: 0, displayNumber: 'Dry run PR' } as never;
-      } else {
+      } else if (!GlobalConfig.get('dryRun')) {
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:438]: revert

File: lib/workers/repository/update/pr/index.ts [hunk 45]
```diff
@@ -496,7 +496,7 @@ export async function ensurePr(
       // istanbul ignore if
       if (GlobalConfig.get('dryRun')) {
         logger.info(`DRY-RUN: Would add comment to PR #${pr.number}`);
-      } else {
+      } else if (!GlobalConfig.get('dryRun')) {
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:499]: revert

File: lib/workers/repository/update/pr/index.ts [hunk 46]
```diff
@@ -516,7 +516,12 @@ export async function ensurePr(
     } else {
       await addAssigneesReviewers(config, pr);
     }
-    logger.debug(`Created ${pr.displayNumber}`);
+    if (
+      GlobalConfig.get('dryRun') !== 'lookup' &&
+      GlobalConfig.get('dryRun') !== 'extract'
+    ) {
+      logger.debug(`Created ${pr.displayNumber}`);
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:523]: Why would you want to print this if dryRun=full?

**Review by rarkins — CHANGES_REQUESTED** [review]
Have you run against a real repo to verify the log output, and can you share some examples?

**Comment by MaronHatoum** [comment]
> Have you run against a real repo to verify the log output, and can you share some examples?

yes, I have logs do you want me to upload the logs here ?

**Comment by MaronHatoum** [comment]
dryRun new options logs:
[dryRun-extract.log](https://github.com/renovatebot/renovate/files/8259938/dryRun-extract.log)
[dryRun-lookup.log](https://github.com/renovatebot/renovate/files/8259939/dryRun-lookup.log)
[dryRun-full.log](https://github.com/renovatebot/renovate/files/8259940/dryRun-full.log)

**Comment by rarkins** [comment]
The extract dry run is missing a log at the end to show the dependencies which were extracted, Eg similar to "packageFiles with updates" after the lookup. You'll need to add this new log message but only log it if dryRun=extract, otherwise it would be unnecessary

File: lib/workers/repository/process/index.ts [hunk 47]
```diff
@@ -103,6 +104,11 @@ export async function extractDependencies(
     logger.debug('No baseBranches');
     const packageFiles = await extract(config);
     addSplit('extract');
+    if (GlobalConfig.get('dryRun') === 'extract') {
+      res.packageFiles = packageFiles;
+      logger.debug({ packageFiles }, 'Extracted dependencies');
```
Comment (rarkins) [inline lib/workers/repository/process/index.ts:109]:
```suggestion
      logger.info({ packageFiles }, 'Extracted dependencies');
```

I think we may as well INFO this for this special case

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Looking good. Will need some sanity testing before merge, e.g. what happens if there's a config error, etc.

**Comment by MaronHatoum** [comment]
> Looking good. Will need some sanity testing before merge, e.g. what happens if there's a config error, etc.

I already ran renovate with all the options:
--dry-run=false 
[dryRun-false.log](https://github.com/renovatebot/renovate/files/8262222/dryRun-false.log)

--dry-run=true
[dryRun-true.log](https://github.com/renovatebot/renovate/files/8262224/dryRun-true.log)

--dry-run=extract
[dryRun-extract.log](https://github.com/renovatebot/renovate/files/8262230/dryRun-extract.log)


--dry-run=lookup
[dryRun-lookup.log](https://github.com/renovatebot/renovate/files/8262228/dryRun-lookup.log)

--dry-run=full
[dryRun-full.log](https://github.com/renovatebot/renovate/files/8262229/dryRun-full.log)

**Comment by rarkins** [comment]
@MaronHatoum the last 4 logs you see in the non dry run mode would be desirable to print for all dry run modes. E.g. timing splits, url stats, "Renovate exiting". Are you able to restore those for the extract and lookup dry run modes?

**Comment by MaronHatoum** [comment]
> @MaronHatoum the last 4 logs you see in the non dry run mode would be desirable to print for all dry run modes. E.g. timing splits, url stats, "Renovate exiting". Are you able to restore those for the extract and lookup dry run modes?

Hi @rarkins 
I added the missing log messages:
extract dry run :
[dryRun-extract.log](https://github.com/renovatebot/renovate/files/8283410/dryRun-extract.log)

lookup dry run:
[dryRun-lookup.log](https://github.com/renovatebot/renovate/files/8283415/dryRun-lookup.log)

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — COMMENTED** [review]
Looks good according to code review, I would like to test it manually before merging

**Review by rarkins — APPROVED** [review]
